pkg/pprof/fix_go_heap_truncated.go

```diff
@@ -0,0 +1,220 @@
+			"inuse_space":
+			return true
+		}
+	}
+	return false
+}
+
+// RepairGoHeapTruncatedStacktraces repairs truncated stack traces
+// in Go heap profiles.
+//
+// Go heap profile has a depth limit of 32 frames, which often
+// renders profiles unreadable, and also increases cardinality
+// of stack traces.
+//
+// The function guesses truncated frames based on neighbors and
+// repairs stack traces if there are high chances that this
+// part is present in the profile. The heuristic is as follows:
+//
+// For each stack trace S taller than 24 frames: if there is another
+// stack trace R taller than 24 frames that overlaps with the given
+// one by at least 16 frames in a row from the top, and has frames
+// above its root, stack S considered truncated, and the missing part
+// is copied from R.
+func RepairGoHeapTruncatedStacktraces(p *profilev1.Profile) {
+	// Group stack traces by bottom (closest to the root) locations.
+	// Typically, there are very few groups (a hundred or two).
+	samples, groups := split(p)
+	// Each group's suffix is then tokenized: each part is shifted by one
+	// location from the previous one (like n-grams).
+	// Tokens are written into the token=>group map, Where the value is the
+	// index of the group with the token found at the furthest position from
+	// the root (across all groups).
+	m := make(map[string]group, len(groups)/2)
+	for i := 0; i < len(groups); i++ {
+		g := groups[i]
+		n := len(groups)
+		if i+1 < len(groups) {
+			n = groups[i+1]
+		}
+		if s := n - g; s < minGroupSize {
+			continue
+		}
+		// We take suffix of the first sample in the group.
+		s := suffix(samples[g].LocationId)
+		// Tokenize the suffix: token position is relative to the stack
+		// trace root: 0 means that the token is the closest to the root.
+		//    TODO: unroll?
+		//    0 : 64 : 192 // No need.
+		//    1 : 56 : 184
+		//    2 : 48 : 176
+		//    3 : 40 : 168
+		//    4 : 32 : 160
+		//    5 : 24 : 152
+		//    6 : 16 : 144
+		//    7 :  8 : 136
+		//    8 :  0 : 128
+		//
+		// We skip the top/right-most token, as it is not needed,
+		// because there can be no more complete stack trace.
+		for j := 1; j <= tokens; j++ {
+			hi := suffixBytesLen - j*tokens
+			lo := hi - tokenBytesLen
+			// By taking a string representation of the slice,
+			// we eliminate the need to hash the token explicitly:
+			// Go map will do it this way or another.
+			k := unsafeString(s[lo:hi])
+			// Current candidate: the group where the token is
+			// located at the furthest position from the root.
+			c, ok := m[k]
+			if !ok || j > c.off {
+				// This group has more complete stack traces:
+				m[k] = group{idx: g, off: j}
+			}
+		}
+	}
+	// Then, for each group, we test, if there is another group with a more
+	// complete suffix, overlapping with the given one by at least one token.
+	// If such stack trace exists, all stack traces of the group are appended
+	// with the missing part.
+	for i := 0; i < len(groups); i++ {
```

Comment:
A note for myself: there is a bug – iterate through the groups in the order of their dependencies